From ed58675e17eeeb037a93f726c2671de61049fc98 Mon Sep 17 00:00:00 2001
From: Viorel Suman <viorel.suman@intel.com>
Date: Wed, 12 Aug 2015 17:22:23 +0300
Subject: [PATCH 07/11] iio: trigger: handle hrtimer start as function of
 trigger state

On "try_reenable" call hrtimer is started regardless of the
hrtimer trigger state thus hrtimer trigger remains started
even if the trigger is explicitly disabled from user space.
The patch fixes this.

Change-Id: Iabaa5144d1dfe164f4ff6b01426deab9894abef6
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-26476
Signed-off-by: Viorel Suman <viorel.suman@intel.com>
Reviewed-on: https://android.intel.com:443/486300
---
 drivers/iio/trigger/iio-trig-hrtimer.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/iio/trigger/iio-trig-hrtimer.c b/drivers/iio/trigger/iio-trig-hrtimer.c
index bc75656..78cd6dd 100644
--- a/drivers/iio/trigger/iio-trig-hrtimer.c
+++ b/drivers/iio/trigger/iio-trig-hrtimer.c
@@ -29,6 +29,7 @@ struct iio_hrtimer_info {
 	unsigned long sampling_frequency;
 	ktime_t period;
 	ktime_t last_event;
+	bool state;
 };
 
 static struct config_item_type iio_hrtimer_type = {
@@ -104,6 +105,7 @@ static int iio_trig_hrtimer_set_state(struct iio_trigger *trig, bool state)
 	struct iio_hrtimer_info *trig_info;
 
 	trig_info = iio_trigger_get_drvdata(trig);
+	trig_info->state = state;
 
 	if (state)
 		hrtimer_start(&trig_info->timer, trig_info->period,
@@ -120,10 +122,11 @@ static int iio_trig_hrtimer_try_reenable(struct iio_trigger *trig)
 	struct iio_hrtimer_info *trig_info;
 
 	trig_info = iio_trigger_get_drvdata(trig);
-	hrtimer_start(&trig_info->timer,
+
+	if (trig_info->state)
+		hrtimer_start(&trig_info->timer,
 			ktime_add(trig_info->last_event, trig_info->period),
 			HRTIMER_MODE_ABS);
-
 	return 0;
 }
 
-- 
1.9.1

